### PR TITLE
Completion: hide private/protected methods/properties

### PR DESCRIPTION
**Pull request by phux**
Just a naive quick fix for hiding private/protected methods & properties, as you mentioned rewriting a new completion engine in #109 and #214

### CODE REVIEW

**Comment by dantleech** [comment]
Nice! Will have a look this evening. Build is failing because of an encrypted secret in travis, but otherwise tests pass.

File: lib/Application/Complete.php [hunk 1]
```diff
@@ -160,11 +160,15 @@ private function populateSuggestions(SymbolContext $symbolInformation, Type $typ
             return $symbolInformation->withIssue(sprintf('Could not find class "%s"', (string) $type));
         }
 
+        $needsPublicAccess = !in_array($symbolInformation->symbol()->name(), ['this', 'self']);
```
Comment (dantleech) [inline lib/Application/Complete.php]: Rename to `$publicOnly`? also `false === in_array` rather than `!`

File: lib/Application/Complete.php [hunk 2]
```diff
@@ -160,11 +160,15 @@ private function populateSuggestions(SymbolContext $symbolInformation, Type $typ
...
         /** @var $method ReflectionMethod */
         foreach ($classReflection->methods() as $method) {
             if ($method->name() === '__construct') {
                 continue;
             }
+            if ($needsPublicAccess && !$method->visibility()->isPublic()) {
```
Comment (dantleech) [inline lib/Application/Complete.php]: `false === $method->visibility()->isPublic()`

File: lib/Application/Complete.php [hunk 3]
```diff
@@ -175,6 +179,9 @@ private function populateSuggestions(SymbolContext $symbolInformation, Type $typ
 
         if ($classReflection instanceof ReflectionClass) {
             foreach ($classReflection->properties() as $property) {
+                if ($needsPublicAccess && !$property->visibility()->isPublic()) {
```
Comment (dantleech) [inline lib/Application/Complete.php]: same

File: lib/Application/Complete.php [hunk 4]
```diff
@@ -160,11 +160,15 @@ private function populateSuggestions(SymbolContext $symbolInformation, Type $typ
             return $symbolInformation->withIssue(sprintf('Could not find class "%s"', (string) $type));
         }
 
+        $needsPublicAccess = !in_array($symbolInformation->symbol()->name(), ['this', 'self']);
         /** @var $method ReflectionMethod */
         foreach ($classReflection->methods() as $method) {
```
Comment (dantleech) [inline lib/Application/Complete.php]: I think the next step (after this PR) would be to add `$classReflection->methods()->visibleTo($symbolInformation->containerType())`

File: lib/Application/Complete.php (hunk 1 above)
Comment (greg0ire) [reply lib/Application/Complete.php]: `in_array` only returns booleans though :P

Comment (greg0ire) [reply lib/Application/Complete.php]: I'd be more worried about not setting its 3rd argument to `true`

Comment (phux) [reply lib/Application/Complete.php]: I agree with greg0ire.

@dantleech if you want or it's the coding style of this project I will change it accordingly.

Comment (dantleech) [reply lib/Application/Complete.php]: > in_array only returns booleans though :P

Strict comparators and boolean keywords are used mostly in the code base - I find it easier to read, but its subjective. Main thing is consistency.

**Comment by dantleech** [comment]
Thanks!

File: lib/Application/Complete.php (hunk 1 above)
Comment (dantleech) [reply lib/Application/Complete.php]: > if you want or it's the coding style of this project I will change it accordingly.

Nah, its fine. Not sure what my algorithm is for the two styles anyway.

**Comment by Taluu** [comment]
Didn't check, but what about parent's protected methods ? And what about static ? :}

Otherwise, just what I've been looking for and complaining about just before updating phpactor :P

**Comment by dantleech** [comment]
Yep, its a naieve implementation - if the variable is `self` or `this` then it will show all visibilities, otherwise only public.

**Comment by Taluu** [comment]
And what if I have `$self`, will it show privates for this one too ?

**Comment by dantleech** [comment]
`self::` and `$this->`

**Pull request merged**